MekHQ/src/mekhq/campaign/Campaign.java

```diff
@@ -3160,7 +3165,37 @@ public void setRetainerEmployerCode(String code) {
         retainerEmployerCode = code;
     }
 
+    private void addInMemoryLogHistory(LogEntry le) {
+        if (inMemoryLogHistory.size() != 0) {
+            long diff = le.getDate().getTime() - inMemoryLogHistory.get(0).getDate().getTime();
+            while ((diff / (1000 * 60 * 60 * 24)) > HistoricalDailyReportDialog.MAX_DAYS_HISTORY) {
```

Comment:
(these are just my general notes, as a random internet person, and not something important to do)

Constantly removing from the head and adding to the tail of a `List<>` data structure is not terribly efficient, due to all of the copies performed (practical viewpoint: who cares, this isn't going to be a huge list, so O(n) is lost in the jitter).

If the data structure was instead a `TreeMap<Integer, List<LogEntry>>` the prune operation becomes easier (although you'd likely want to ditch the `LogEntry` in favor of a new `DailyLog` object which you could reuse and append log entries to).
```java
// assumes: TreeMap<Long, xxx> inMemoryLogHistory;
Long currentTime = le.getDate().getTime();
Long pruneBefore = currentTime - (HistoricalDailyReportDialog.MAX_DAYS_HISTORY * 1000 * 60 * 60 * 24);
inMemoryLogHistory.headMap(pruneBefore).clear();
xxx dailyLog = inMemoryLogHistory.get(currentTime);
if (dailyLog == null) {
  // ...add a new daily log
}
else {
  // ...append to the daily log
}
```
Taking it one step further you'd abstract this out entirely into a `CampaignReport` class which would be used for all campaign reporting and campaign history, hiding the machinery from `Campaign` (promoting reuse while decreasing internal coupling). This would also promote further enhancements to reporting and historical logging.

Again, I'm just some internet person who dropped by: zero requirement to follow any of this.

Reply:
From a pure computer science point of view, you want a Queue. You probably want a LinkedList, as that implements the Queue interface.